bytecomp/matching.ml

```diff
@@ -732,7 +732,7 @@ end = struct
                      because we cannot express
                         (K (p1, .., pn) | K (q1, .. qn))
                      as (p1 .. pn | q1 .. qn) *)
-                  filter_rec [ p1 :: ps; p2 :: ps ] @ rem
+                  filter_rec ((p1 :: ps) :: (p2 :: ps) :: rem)
```

Comment:
I spent a long time looking at this thinking "but these two lines are equivalent!" before asking a colleague for help.
Turns out that while `[ p1 :: ps; p2 :: ps ] @ rem` and `(p1 :: ps) :: (p2 :: ps) :: rem` are equivalent `foo [ bar ] @ baz` and `foo ([ bar ] @ baz)` are not!
Turns out that this diff was putting emphasis on the unimportant parts!

Reply:
The diff (1) fixes the bug and (2) changes the syntax to make the bug impossible. It's optimized for the robustness of the output, rather than readability of the diff.

Reply:
It was not a critique!

However, since we're on that topic: you can have the best of both worlds, simply split the commit in two! One that fixes the bug, and a follow up one which changes the syntax.

Reply:
(I'm trying to answer comments and hear research talks at the same talk, which is a terrible idea and might make me sound more dry than I intend :-)